core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

```diff
@@ -719,7 +721,27 @@ object SparkSubmit {
       printWarning("Subclasses of scala.App may not work correctly. Use a main() method instead.")
     }
 
-    val mainMethod = mainClass.getMethod("main", new Array[String](0).getClass)
+    val isSparkApp = mainClass.getMethods().filter(_.getName() == "sparkMain").length > 0
```

Comment:
Can you use `classOf[SparkApp].isAssignableFrom(mainClass)` here? Or that doesn't work?